crates/metadata-store/src/rocksdb/server.rs

```diff
@@ -0,0 +1,344 @@
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use bytes::Bytes;
+use bytestring::ByteString;
+use restate_core::network::{MessageHandler, NetworkSender};
+use restate_core::{cancellation_watcher, task_center, TaskKind};
+use restate_node_protocol::metadata_store::{
+    Delete, Get, GetVersion, MetadataStoreRequest, MetadataStoreResponse, Put, RequestKind,
+    ResponseKind,
+};
+use restate_node_protocol::{metadata_store, MessageEnvelope};
+use restate_types::errors::GenericError;
+use restate_types::metadata_store::{Precondition, VersionedValue};
+use restate_types::{GenerationalNodeId, NodeId, Version};
+use rocksdb::{ColumnFamily, Options, WriteOptions, DB};
+use serde::de::DeserializeOwned;
+use serde::Serialize;
+use std::path::Path;
+use tokio::sync::mpsc;
+use tracing::{debug, trace};
+
+// todo: Introduce bound if this channel should grow unboundedly
+type CommandSender = mpsc::UnboundedSender<(GenerationalNodeId, MetadataStoreRequest)>;
+type CommandReceiver = mpsc::UnboundedReceiver<(GenerationalNodeId, MetadataStoreRequest)>;
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("storage error: {0}")]
+    Storage(#[from] rocksdb::Error),
+    #[error("violated precondition: {0}")]
+    PreconditionViolation(String),
+    #[error("codec error: {0}")]
+    Codec(GenericError),
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum BuildError {
+    #[error("failed opening rocksdb: {0}")]
+    RocksDB(#[from] rocksdb::Error),
+}
+
+impl Error {
+    fn kv_pair_exists() -> Self {
+        Error::PreconditionViolation("key-value pair already exists".to_owned())
+    }
+
+    fn version_mismatch(expected: &Version, actual: Option<&Version>) -> Self {
+        Error::PreconditionViolation(format!(
+            "Expected version '{}' but found version '{:?}'",
+            expected, actual
+        ))
+    }
+}
+
+impl From<Error> for metadata_store::Error {
+    fn from(err: Error) -> Self {
+        match err {
+            Error::PreconditionViolation(msg) => metadata_store::Error::PreconditionViolation(msg),
+            err => metadata_store::Error::Internal(err.to_string()),
+        }
+    }
+}
+
+type Result<T> = std::result::Result<T, Error>;
+
+const KV_PAIRS: &str = "kv_pairs";
+
+/// Single node metadata store which stores the key value pairs in RocksDB.
+///
+/// In order to avoid issues arising from concurrency, we run the metadata
+/// store in a single thread.
+pub struct RocksDBMetadataStore<N> {
+    db: DB,
+    write_opts: WriteOptions,
+    command_rx: CommandReceiver,
+    networking: N,
+
+    // for creating other senders
+    command_tx: CommandSender,
+}
+
+impl<N> RocksDBMetadataStore<N>
+where
+    N: NetworkSender + Send + 'static,
+{
+    pub fn new(networking: N, path: impl AsRef<Path>) -> std::result::Result<Self, BuildError> {
+        let (command_tx, command_rx) = mpsc::unbounded_channel();
```

Comment:
I personally default to bounded channels, but I wouldn't be too fussy about this at this stage.

Reply:
Better be safe than sorry. Will change it.